### PR TITLE
Sih fw

### PR DESCRIPTION
**Pull request by romain-chiap**
Hi everyone,

I'm excited to release a fixed wing model for the simulator in hardware. The aerodynamic model is based on the Thesis by 

> Waqas Khan, supervised by Meyer Nahon "Dynamics modeling of agile fixed-wing unmanned aerial vehicles." McGill University, PhD thesis, 2016. Sections 3.1, 3.2, and 3.3

I implemented it in the class AeroSeg, which is capable of
- full 360 deg angle of attack lift, drag, and pitching moment.
- wings with aspect ratio from 0.1666 up to 6, (gliders would perform poorly for instance).
- control surface (flap) deflection up to 70 degrees.
- unlimited flap chord, (the elevator could take the entire tailplane).
- stall angle function of the flap deflection.
- effect of the flap deflection on lift, drag, and pitching moment.
- any dihedral angle, the fin is modeled as a wing with dihedral angle of 90 deg.
- slipstream velocity (velocity from the propeller) using momentum theory.

This class is meant to model different vehicle easily by specifying the geometry of their wings and tail surfaces.

Notes:
- I need to takeoff in Manual mode with full throttle.
- The floor is a simple model where the vehicle slides on it, it is also automatically replaced flat for the next takeoff which throws off the EKF when landing.

Here are the full 360 deg aerodynamic coefficient for a wing with span 6m, chord 1m and flap chord 0.25m.
aoa_full_smooth_AR6

https://user-images.githubusercontent.com/17630348/127492813-4e1efe62-4bbe-44fd-9a95-3f498079bdb9.mp4

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Woah! This is super cool! Now we need VTOL 8)

**Comment by romain-chiap** [comment]
> Woah! This is super cool! Now we need VTOL 8)

Yes, this is why the aerodynamic class is designed to handle the 360 deg angle of attack!

**Comment by Jaeyoung-Lim** [comment]
Really awesome!

File: src/modules/sih/sih.hpp [hunk 1]
```diff
@@ -158,10 +204,20 @@ class Sih : public ModuleBase<Sih>, public ModuleParams, public px4::ScheduledWo
 	matrix::Quatf       _q;             // quaternion attitude
 	matrix::Dcmf        _C_IB;          // body to inertial transformation
 	matrix::Vector3f    _w_B;           // body rates in body frame [rad/s]
-	matrix::Quatf       _q_dot;         // quaternion differential
+	matrix::Quatf       _dq;            // quaternion differential
 	matrix::Vector3f    _w_B_dot;       // body rates differential
-	float       _u[NB_MOTORS];  // thruster signals
+	float       _u[NB_MOTORS];          // thruster signals
+
+	enum Vtype {MC, FW}; 	// vehicle type
```
Comment (bkueng) [inline src/modules/sih/sih.hpp:211]:
```suggestion
	enum class VehicleType {MC, FW};
```

File: src/modules/sih/sih.hpp [hunk 2]
```diff
@@ -113,6 +136,10 @@ class Sih : public ModuleBase<Sih>, public ModuleParams, public px4::ScheduledWo
 	vehicle_global_position_s			_gpos_gt{};
 	uORB::Publication<vehicle_global_position_s>	_gpos_gt_pub{ORB_ID(vehicle_global_position_groundtruth)};
 
+	// airspeed
+	airspeed_s  						_airspeed{};
```
Comment (bkueng) [inline src/modules/sih/sih.hpp:140]: This is filled in entirely in `send_airspeed`, can you keep it locally in there?

File: src/modules/sih/aero.hpp [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+		aoa_coeff(alpha, sqrtf(vxz2), def);
+		Fa = C_BS * (0.5f * rho * vxz2 * span * mac) * matrix::Vector3f(CL * sinf(alpha) - CD * cosf(alpha),
+				0.0f,
+				-CL * cosf(alpha) - CD * sinf(alpha));
+		Ma = C_BS * (0.5f * rho * vxz2 * span * mac * mac) * matrix::Vector3f(0.0f, CM,
+				0.0f) + p_B % Fa; 	// computed at vehicle CM
+	}
+
+	// return the air density at current altitude, must be called after update_aero()
+	float get_rho() {	return rho; 	}
```
Comment (bkueng) [inline src/modules/sih/aero.hpp:245]:
```suggestion
	float get_rho() const { return rho; }
```

File: src/modules/sih/aero.hpp [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+	float pressure; 	// pressure in Pa at current altitude
+	float temperature;	// temperature in K at current altitude
+	float prop_radius;	// propeller radius [m], used to create the slipstream
+	float v_slipstream;	// slipstream velocity [m/s], computed from momentum theory
+
+public:
+
+	matrix::Vector3f Fa;	// aerodynamic force
+	matrix::Vector3f Ma;	// aerodynamic moment computed at CM directly
```
Comment (bkueng) [inline src/modules/sih/aero.hpp:140]: Can you keep this private and use getters?

File: src/modules/sih/aero.hpp [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+
+	static constexpr const float KV = M_PI_F; 	// total vortex lift parameter
+	static constexpr float CD0 = 0.04f; 		// no lift drag coefficient
+	static constexpr float CD90 = 1.98f; 		// 90 deg angle of attack drag coefficient
+	static constexpr float AF_DOT_MAX = M_PI_F / 2.0f;
+	static constexpr float ALPHA_BLEND = M_PI_F / 18.0f; 	// 10 degrees
+
+	// here we make the distinction of the plate (i.e. wing, or tailplane, or fin) and the segment
+	// the segment can be a portion of the wing, but the aspect ratio (AR) of the wing needs to be used
+	float alpha; 		// angle of attack [rad]
```
Comment (bkueng) [inline src/modules/sih/aero.hpp:99]:
```suggestion
	float _alpha; 		// angle of attack [rad]
```
For the rest as well

File: src/modules/sih/sih.cpp [hunk 6]
```diff
@@ -430,6 +516,22 @@ void Sih::publish_sih()
 	_gpos_gt_pub.publish(_gpos_gt);
 }
 
+// quaternion exponential as defined in [3]
+Quatf Sih::expq(matrix::Vector3f u)
```
Comment (bkueng) [inline src/modules/sih/sih.cpp:520]:
```suggestion
Quatf Sih::expq(const matrix::Vector3f& u)
```

File: src/modules/sih/sih.cpp [hunk 7]
```diff
@@ -430,6 +516,22 @@ void Sih::publish_sih()
...
+{
+	float u_norm = u.norm();
+	Vector3f v;
+
+	if (fabsf(u_norm) < 1.0e-6f) { 	// error will be smaller than 1e-18
```
Comment (bkueng) [inline src/modules/sih/sih.cpp:525]:
```suggestion
	if (u_norm < 1.0e-6f) { 	// error will be smaller than 1e-18
```

File: src/modules/sih/sih.cpp [hunk 8]
```diff
@@ -430,6 +516,22 @@ void Sih::publish_sih()
...
+		v = (1.0f - u_norm * u_norm / 6.0f) * u; 	// first taylor serie term of sin(x)/x
+
+	} else {
+		v = sinf(u_norm) / u_norm * u;
+	}
+
+	return Quatf(cosf(u.norm()), v(0), v(1), v(2));
```
Comment (bkueng) [inline src/modules/sih/sih.cpp:532]:
```suggestion
	return Quatf(cosf(u_norm), v(0), v(1), v(2));
```

File: ROMFS/px4fmu_common/init.d/airframes/1101_rc_plane_sih.hil [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+# @board px4_fmu-v2 exclude
+#
+
+. ${R}etc/init.d/rc.fw_defaults
+
+set MIXER AERT
+set PWM_OUT 1234
+
+# set SYS_HITL to 2 to start the SIH and avoid sensors startup
+param set SYS_HITL 2
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/airframes/1101_rc_plane_sih.hil:19]:
```suggestion
param set-default SYS_HITL 2
```

**Review by bkueng — COMMENTED** [review]
Cool!

File: src/modules/sih/sih.cpp (hunk 6 above)
Comment (romain-chiap) [reply src/modules/sih/sih.cpp:520]: is it about saving memory by avoiding to reallocate?

Comment (bkueng) [reply src/modules/sih/sih.cpp]: It's to avoid having to copy the vector when calling the method.

File: src/modules/sih/sih_params.c [hunk 10]
```diff
@@ -435,3 +435,13 @@ PARAM_DEFINE_FLOAT(SIH_DISTSNSR_OVR, -1.0f);
  * @group Simulation In Hardware
  */
 PARAM_DEFINE_FLOAT(SIH_T_TAU, 0.05f);
+
+/**
+ * Vehicle type (0 for MC, 1 for FW)
```
Comment (bkueng) [inline src/modules/sih/sih_params.c:440]:
```suggestion
 * Vehicle type
```

**Review by bkueng — DISMISSED** [review]
Can you update the jmavsim submodule commit to master?

**Review by bkueng — APPROVED** [review]

**Comment by hamishwillee** [comment]
@bkueng The airframe updates didn't propagate to PX4 user guide (this was merged 6 days ago and last update was 9 days ago). Similarly, the parameters did not update either (can't see SIH_VEHICLE_TYPE in https://github.com/PX4/PX4-user_guide/blob/master/en/advanced_config/parameter_reference.md#simulation-in-hardware)

Can you identify why/retrigger update?

**Comment by bkueng** [comment]
This should run on Jenkins and I don't have insights into that. @dagar?

**Pull request merged**